File: js/src/providers/mistral.ts
import { modelSpec, ModelSpec } from "../registry.js";
import { Provider, ProviderOptions, TierLimit } from "./base.js";

/** Mistral AI — OpenAI-compatible, free "Experiment" tier.
 * Free tier (verified 2026-06): 2 req/min, 500K tokens/min, 1B tokens/month.
 * Model ids self-correct via live /models discovery. */
export class Mistral extends Provider {
  static providerName = "mistral";
  static baseUrl = "https://api.mistral.ai/v1";
  static tiers: Record<string, TierLimit> = {
    free: { rpm: 2, rpd: null },
  };
  static defaultModels: ModelSpec[] = [
    modelSpec("mistral-small-latest", ["chat", "large", "tools"], 32000),
    modelSpec("mistral-large-latest", ["chat", "large", "tools"], 128000),
  ];

  constructor(keys: string | string[], opts: ProviderOptions = {}) {
    super(keys, { discover: true, ...opts });
  }
}

File: src/freelm/providers/mistral.py
"""Mistral AI (https://mistral.ai) — OpenAI-compatible, free "Experiment" tier."""
from __future__ import annotations

from typing import Any, Dict

from ..registry import ModelSpec
from .base import Provider


class Mistral(Provider):
    name = "mistral"
    base_url = "https://api.mistral.ai/v1"

    # Free "Experiment" tier (verified 2026-06): 2 req/min, 500K tokens/min,
    # 1B tokens/month. The low RPM is real — add a paid tier or override rpm.
    TIERS: Dict[str, Dict[str, Any]] = {
        "free": {"rpm": 2, "rpd": None},
    }

    DEFAULT_MODELS = [
        ModelSpec("mistral-small-latest", ("chat", "large", "tools"), ctx=32000),
        ModelSpec("open-mistral-nemo", ("chat", "small", "fast"), ctx=128000),
    ]
